### PR TITLE
AirLoopHVAC:UnitaryHeatCool:VAVChangeoverBypass with Fan:SystemModel crashes

### LINKED ISSUE #10252: AirLoopHVAC:UnitaryHeatCool:VAVChangeoverBypass fails with Fan:SystemModel component
Issue overview
--------------
Test file shows a crash when using Fan:SystemModel in CBVAV simulation.

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version)
 - Version of EnergyPlus (if using an intermediate build, include SHA)
 - Unmethours link or helpdesk ticket number

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added ([in-segFault.idf.txt](https://github.com/NREL/EnergyPlus/files/12751246/in-segFault.idf.txt))
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by rraustad**
Pull request overview
---------------------
 - Fixes #10252
 - HVACUnitaryCBVAV model incorrectly accesses Fan:SystemModel index

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/HVACUnitaryBypassVAV.cc [hunk 1]
```diff
@@ -616,8 +616,8 @@ namespace HVACUnitaryBypassVAV {
                 thisCBVAV.FanVolFlow = 9999.0;
             } else {
                 if (thisCBVAV.FanType == DataHVACGlobals::FanType::System) {
-                    thisCBVAV.FanIndex = HVACFan::getFanObjectVectorIndex(state, thisCBVAV.FanName);
                     state.dataHVACFan->fanObjs.emplace_back(new HVACFan::FanSystem(state, thisCBVAV.FanName)); // call constructor
+                    thisCBVAV.FanIndex = HVACFan::getFanObjectVectorIndex(state, thisCBVAV.FanName);
                     thisCBVAV.FanInletNodeNum = state.dataHVACFan->fanObjs[thisCBVAV.FanIndex]->inletNodeNum;
                     fanOutletNode = state.dataHVACFan->fanObjs[thisCBVAV.FanIndex]->outletNodeNum;
                     thisCBVAV.FanVolFlow = state.dataHVACFan->fanObjs[thisCBVAV.FanIndex]->designAirVolFlowRate;
```
Comment (rraustad) [inline src/EnergyPlus/HVACUnitaryBypassVAV.cc:623]: Conversion of model getInput to eliminate mining functions uses getFanIndex first for legacy fans and emplace_back THEN getFanIndex for Fan:SystemModel.

Comment (jmarrec) [reply src/EnergyPlus/HVACUnitaryBypassVAV.cc:623]: `thisCBVAV.FanIndex = state.dataHVACFan->fanObjs.size() - 1;` should work and be less computationally intensive unless I'm missing something?

Comment (rraustad) [reply src/EnergyPlus/HVACUnitaryBypassVAV.cc:623]: I was hoping to reduce all the mining functions to a single getIndex for each component type so all getInput code would look the same. But your right, here, right after the emplace_back, the index is as you say, size - 1.

Comment (Myoldmopar) [reply src/EnergyPlus/HVACUnitaryBypassVAV.cc:623]: Sure, this could do for a cleanup, but not today.  Thanks @rraustad

**Review by Myoldmopar — APPROVED** [review]

**Comment by Myoldmopar** [comment]
One unrelated diff, this is good to go.  Locally I checked out latest develop and pulled in both this branch and the #10254 branch, simulating what will happen when I merge both of them down.  I am testing in that state right now and if it passes, I'll merge both of these in and directly tag RC3.

**Pull request merged**